include/matx/transforms/pwelch.h

```diff
@@ -0,0 +1,83 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+/////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+namespace matx
+{
+  /**
+   *  Estimate the power spectral density of a 1D tensor using Welch's method [1].
+   *
+   * @param Pxx
+   *   Output power spectral density of time domain signal x
+   * @param x
+   *   Input time domain signal x
+   * @param nperseg
+   *   Length of each segment
+   * @param stream
+   *   cuda Stream to execute on
+   * @param noverlap
+   *   Number of points to overlap between segments.  Defaults to 0
+   * @param nfft
+   *   Length of FFT used per segment.  nfft >= nperseg.  Defaults to nfft = nperseg
+   *
+   * [1] P. Welch, "The use of fast Fourier transform for the estimation of power spectra: A method based on time averaging over short, modified periodograms," in IEEE Transactions on Audio and Electroacoustics, vol. 15, no. 2, pp. 70-73, June 1967, doi: 10.1109/TAU.1967.1161901.
+   *
+   */
+  template <typename PxxType, typename xType>
+    __MATX_INLINE__ void pwelch_impl(PxxType Pxx, xType x, int nperseg, int noverlap, int nfft, cudaStream_t stream=0)
+    {
+      MATX_NVTX_START("", matx::MATX_NVTX_LOG_API)
+
+      MATX_ASSERT_STR(Pxx.Rank() == x.Rank(), matxInvalidDim, "pwelch:  Pxx rank must be the same as x rank");
+      MATX_ASSERT_STR(nfft >= nperseg, matxInvalidDim, "pwelch:  nfft must be >= nperseg");
+      MATX_ASSERT_STR((noverlap >= 0) && (noverlap < nperseg), matxInvalidDim, "pwelch:  Must have 0 <= noverlap < nperseg");
+
+      // Create overlapping view
+      auto x_with_overlaps = x.OverlapView({nfft}, {noverlap});
+
+      // Create temporary space for fft outputs
+      index_t batches = x_with_overlaps.Shape()[0];
+      auto X_with_overlaps = make_tensor<cuda::std::complex<typename PxxType::scalar_type>>({batches,static_cast<index_t>(nfft)});
+
+      (X_with_overlaps = fft(x_with_overlaps,nfft)).run(stream);
+
+      // Compute magnitude squared in-place
+      (X_with_overlaps = conj(X_with_overlaps) * X_with_overlaps).run(stream);
+      auto mag_sq_X_with_overlaps = X_with_overlaps.RealView();
+
+      // Perform the reduction across 'batches' rows and normalize
+      float norm_factor = 1.f / static_cast<float>(batches);
```

Comment:
May not be necessary, but `norm_factor` could be cast to `inner_type_t` so it's a double * double or float * float.